src/rust/engine/options/src/args.rs

```diff
@@ -143,25 +146,73 @@ impl Args {
     }
 }
 
+pub(crate) struct ArgsTracker {
+    unconsumed_args: Mutex<HashSet<Arg>>,
+}
+
+impl ArgsTracker {
+    fn new(args: &Args) -> Self {
+        Self {
+            unconsumed_args: Mutex::new(args.args.clone().into_iter().collect()),
+        }
+    }
+
+    fn consume_arg(&self, arg: &Arg) {
+        self.unconsumed_args.lock().remove(arg);
+    }
+
+    pub fn get_unconsumed_flags(&self) -> Vec<String> {
+        let mut ret: Vec<String> = self
+            .unconsumed_args
+            .lock()
+            .iter()
+            .map(|arg| arg.flag.clone())
+            .collect();
+        ret.sort(); // For stability in tests and when reporting unconsumed args.
+        ret
+    }
+}
+
 pub(crate) struct ArgsReader {
     args: Args,
-    #[allow(dead_code)]
     fromfile_expander: FromfileExpander,
+    tracker: Arc<ArgsTracker>,
 }
 
 impl ArgsReader {
     pub fn new(args: Args, fromfile_expander: FromfileExpander) -> Self {
+        let tracker = Arc::new(ArgsTracker::new(&args));
         Self {
             args,
             fromfile_expander,
+            tracker,
         }
     }
 
-    #[allow(dead_code)]
     pub fn get_passthrough_args(&self) -> Option<&Vec<String>> {
         self.args.passthrough_args.as_ref()
     }
 
+    pub fn get_tracker(&self) -> Arc<ArgsTracker> {
+        self.tracker.clone()
+    }
+
+    fn matches(&self, arg: &Arg, id: &OptionId) -> bool {
```

Comment:
There is no such concept today, but I can rename just in case.

Reply:
Although that might be more clunky than it's worth. In any case, if you have an id in hand and there is a match, even if you don't do anything with that match, it's enough for us to know that the option is "known"/"registered" (otherwise we wouldn't know how to construct the id)

Reply:
On balance, decided not to rename.

Reply:
Sounds good. 

> it's enough for us to know that the option is "known"/"registered"

Ah, right, yes. 👍